runtime/compiler/z/codegen/J9S390PrivateLinkage.cpp

```diff
@@ -2823,6 +2823,73 @@ void TR::J9S390JNILinkage::checkException(TR::Node * callNode,
    generateS390LabelInstruction(self()->cg(), TR::InstOpCode::LABEL, callNode, exceptionRestartLabel);
    }
 
+/**
+ * \brief JNI return value processing:
+ * 1) Unwrap return value for objects or
+ * 2) Enforce a return value of 0 or 1 for boolean return type.
+*/
+void
+TR::J9S390JNILinkage::processJNIReturnValue(TR::Node * callNode,
+                                            TR::CodeGenerator* cg,
+                                            TR::Register* javaReturnRegister,
+                                            TR::RegisterDependencyConditions* deps)
+   {
+   auto resolvedMethod = callNode->getSymbol()->castToResolvedMethodSymbol()->getResolvedMethod();
+   auto returnType = resolvedMethod->returnType();
+   bool isUnwrapAddressReturnValue = !((TR_J9VMBase *)fe())->jniDoNotWrapObjects(resolvedMethod)
+                                     && (returnType == TR::Address)
+                                     && (javaReturnRegister!= NULL)
+                                     && (javaReturnRegister->getKind() == TR_GPR);
+
+   TR::LabelSymbol *cFlowRegionStart = NULL, *cFlowRegionEnd = NULL;
+
+   if (isUnwrapAddressReturnValue)
+      {
+      cFlowRegionStart = generateLabelSymbol(cg);
+      cFlowRegionEnd = generateLabelSymbol(cg);
+
+      generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, callNode, cFlowRegionStart);
+      cFlowRegionStart->setStartInternalControlFlow();
+      generateS390CompareAndBranchInstruction(cg, TR::InstOpCode::getCmpOpCode(), callNode, javaReturnRegister, (signed char)0, TR::InstOpCode::COND_BE, cFlowRegionEnd);
+      generateRXInstruction(cg, TR::InstOpCode::getLoadOpCode(), callNode, javaReturnRegister,
+                 generateS390MemoryReference(javaReturnRegister, 0, cg));
+
+      int32_t regPos = deps->searchPostConditionRegisterPos(javaReturnRegister);
+      TR::RealRegister::RegNum realReg = deps->getPostConditions()->getRegisterDependency(regPos)->getRealRegister();
+      TR::RegisterDependencyConditions * postDeps = new (trHeapMemory()) TR::RegisterDependencyConditions(0, 1, cg);
+      postDeps->addPostCondition(javaReturnRegister, realReg);
+
+      generateS390LabelInstruction(cg, TR::InstOpCode::LABEL, callNode, cFlowRegionEnd, postDeps);
+      cFlowRegionEnd->setEndInternalControlFlow();
+      }
+   else if ((returnType == TR::Int8) && comp()->getSymRefTab()->isReturnTypeBool(callNode->getSymbolReference()))
+      {
+      if (cg->getS390ProcessorInfo()->supportsArch(TR_S390ProcessorInfo::TR_z13))
+         {
+         generateRIInstruction(cg, TR::InstOpCode::getCmpHalfWordImmOpCode(), callNode, javaReturnRegister, (int32_t) 0);
```

Comment:
I don't think we need a `(int32_t)` cast here.

Reply:
Fixed.